### PR TITLE
feat: store deployment key for deployed decisions

### PR DESCRIPTION
**Pull request by jfriedenstab**
## Description
Store the deployment key as a new property for deployed decisions and add the possibility to find a decision by decision id and deployment key. This is a prerequisite to be able to properly resolve called decisions when the new `deployment` binding type is used (see #19710).

## Related issues
part of #19709 (processes were already handled with #20498; forms will follow in another PR)

### CODE REVIEW

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbDecisionState.java [hunk 1]
```diff
@@ -348,7 +385,6 @@ public void storeDecisionRecord(final DecisionRecord record) {
     dbPersistedDecision.wrap(record);
     decisionsByKey.upsert(tenantAwareDecisionKey, dbPersistedDecision);
 
-    dbDecisionKey.wrapLong(record.getDecisionKey());
```
Comment (jfriedenstab) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbDecisionState.java:351]: Decision key is already wrapped in line 347.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/CreateDeploymentMultiplePartitionsTest.java [hunk 2]
```diff
@@ -534,6 +534,33 @@ public void shouldWriteDrgAndDecisionCreatedEventsWithSameKeys() {
...
+    // then
+    ENGINE.forEachPartition(
+        partitionId -> {
+          final var record =
+              RecordingExporter.decisionRecords()
+                  .withPartitionId(partitionId)
+                  .withIntents(DecisionIntent.CREATED)
+                  .withDecisionId(decisionId)
+                  .limit(1)
+                  .getFirst();
```
Comment (korthout) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/CreateDeploymentMultiplePartitionsTest.java:555]: 🔧 Nit: no need to limit to 1 if we're getting the first. `getFirst()` would already short circuit if an entry is available.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/mutable/MutableDecisionState.java [hunk 3]
```diff
@@ -28,6 +28,14 @@ public interface MutableDecisionState extends DecisionState {
    */
   void storeDecisionRequirements(DecisionRequirementsRecord record);
 
+  /**
+   * Store a reference to the decision key in the state, using the decision id and deployment key as
+   * composite database key.
+   *
+   * @param record the record of the decision
+   */
```
Comment (korthout) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/mutable/MutableDecisionState.java:36]: 👍 I like the docs

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/deployment/DecisionStateTest.java [hunk 4]
```diff
@@ -167,6 +180,92 @@ void shouldFindDeployedDecisionByKey() {
         .isEqualTo(decisionRecord2.getDecisionKey());
   }
 
+  @DisplayName("should find deployed decision by ID and deployment key")
+  @Test
+  void shouldFindDeployedDecisionByIdAndDeploymentKey() {
```
Comment (korthout) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/deployment/DecisionStateTest.java:185]: 👍 Good test

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/deployment/DecisionStateMultiTenantTest.java [hunk 5]
```diff
@@ -105,6 +105,47 @@ void shouldPutDecisionForDifferentTenants() {
...
+    final var drgKey = 123L;
+    final var drgId = "drgId";
+    final var decisionKey = 456L;
+    final var decisionId = "decisionId";
+    final int version = 1;
+    final var tenant1 = "tenant1";
+    final var tenant2 = "tenant2";
+    final var deploymentKey = 789L;
+    final var drg1 = sampleDecisionRequirementsRecord(tenant1, drgKey, drgId, version);
+    final var drg2 = sampleDecisionRequirementsRecord(tenant2, drgKey, drgId, version);
```
Comment (korthout) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/deployment/DecisionStateMultiTenantTest.java:120]: 💭 This test is a bit strange. It considers 2 DRGs belonging to different tenants, but they have the same key (`drgKey`). This already is the case in the test above (where this is likely copied from), so let's just accept it. It's not what this test focusses on, but just seems a bit strange.

**Review by korthout — APPROVED** [review]
Awesome @jfriedenstab 🚀 

👍 LGTM

**Review by korthout — CHANGES_REQUESTED** [review]
I see now that there's something missing

❌ Please ensure we remove decisions from the new column family on deletion

**Comment by jfriedenstab** [comment]
Thanks for the review!

> ❌ Please ensure we remove decisions from the new column family on deletion

This is actually done here:
https://github.com/camunda/camunda/blob/eba24c9386eaff49e48608dd4b1cf60aec2540a3/zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbDecisionState.java#L461
(and also covered by "should not find decision after it has been deleted" in `DecisionStateTest`).

Or did I miss anything?

**Comment by korthout** [comment]
Oops. I missed it. Sorry @jfriedenstab

**Review by korthout — APPROVED** [review]
LGTM

**Pull request merged**